library/src/main/java/com/novoda/downloadmanager/DownloadManager.java

```diff
@@ -123,48 +125,60 @@ public void removeDownloadBatchCallback(DownloadBatchCallback downloadBatchCallb
         }
     }
 
+    @WorkerThread
     @Override
-    public void getAllDownloadBatchStatuses(AllBatchStatusesCallback callback) {
-        if (downloadService == null) {
-            executor.submit(
-                    WaitForLockRunnable.waitFor(waitForDownloadService)
-                            .thenPerform(() -> executeGetAllDownloadBatchStatuses(callback))
-            );
-        } else {
-            executeGetAllDownloadBatchStatuses(callback);
-        }
+    public List<DownloadBatchStatus> getAllDownloadBatchStatuses() {
+        return WaitForDownloadServiceThenPerform.<List<DownloadBatchStatus>>waitFor(downloadService, waitForDownloadService)
+                .thenPerform(this::executeGetAllDownloadBatchStatuses);
     }
 
-    private void executeGetAllDownloadBatchStatuses(AllBatchStatusesCallback callback) {
+    private List<DownloadBatchStatus> executeGetAllDownloadBatchStatuses() {
         List<DownloadBatchStatus> downloadBatchStatuses = new ArrayList<>(downloadBatchMap.size());
 
         for (DownloadBatch downloadBatch : downloadBatchMap.values()) {
             downloadBatchStatuses.add(downloadBatch.status());
         }
+        return downloadBatchStatuses;
+    }
 
-        callbackHandler.post(() -> callback.onReceived(downloadBatchStatuses));
+    @Override
+    public void getAllDownloadBatchStatuses(AllBatchStatusesCallback callback) {
+        executor.submit((Runnable) () -> WaitForDownloadServiceThenPerform.<Void>waitFor(downloadService, waitForDownloadService)
+                .thenPerform(() -> {
+                    List<DownloadBatchStatus> downloadBatchStatuses = executeGetAllDownloadBatchStatuses();
+                    callbackHandler.post(() -> callback.onReceived(downloadBatchStatuses));
+                    return null;
+                }));
     }
 
+    @Nullable
+    @WorkerThread
     @Override
-    public void getDownloadStatusWithMatching(DownloadFileId downloadFileId, DownloadFileStatusCallback callback) {
-        if (downloadService == null) {
-            executor.submit(
-                    WaitForLockRunnable.waitFor(waitForDownloadService)
-                            .thenPerform(() -> executeFirstLocalPathForDownloadWithMatching(downloadFileId, callback))
-            );
-        } else {
-            executeFirstLocalPathForDownloadWithMatching(downloadFileId, callback);
-        }
+    public DownloadFileStatus getDownloadStatusWithMatching(DownloadFileId downloadFileId) {
+        return WaitForDownloadServiceThenPerform.<DownloadFileStatus>waitFor(downloadService, waitForDownloadService)
```

Comment:
How does it look adding a static import for WaitForDownloadServiceThenPerform.waitFor()?

Reply:
I don't think you can add a static because of the generic type.